broker/src/main/java/org/apache/rocketmq/broker/client/ConsumerGroupInfo.java

```diff
@@ -136,7 +137,7 @@ public boolean updateChannel(final ClientChannelInfo infoNew, ConsumeType consum
                     this.groupName,
                     infoOld.toString(),
                     infoNew.toString());
-                this.channelInfoTable.put(infoNew.getChannel(), infoNew);
+                this.channelInfoTable.put(infoNew.getRemotingChannel(), infoNew);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **124, 125, 135, and 140**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread's `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).